crates/admin/src/cluster_controller/replicated_loglet_controller.rs

```diff
@@ -0,0 +1,656 @@
+            replication = self.config.replication.num_copies(),
+            "Deciding on next cluster plan"
+        );
+
+        if have_workers && have_log_servers {
+            // First, create loglet segments for any logs that don't exist in the plan
+            let mut updated_plan = scheduling_plan.clone().into_builder();
+            let mut effects = vec![];
+
+            for log_id in cluster_state.logs.iter() {
+                match scheduling_plan.loglet_config(log_id) {
+                    None => {
+                        let node_set = NodeSet::from_nodes(
+                            cluster_state
+                                .healthy_log_servers
+                                .keys()
+                                .cloned()
+                                .choose_multiple(
+                                    &mut rand::thread_rng(),
+                                    self.config.replication.num_copies() as usize,
+                                ),
+                        );
+                        assert_eq!(
+                            node_set.len(),
+                            self.config.replication.num_copies() as usize
+                        );
+
+                        let sequencer = cluster_state
+                            .healthy_workers
+                            .values()
+                            .choose(&mut rand::thread_rng())
+                            .cloned()
+                            .expect("we have at least one healthy worker");
+
+                        let loglet_id = rand::thread_rng().next_u64(); // should we make this a proper Restate ID?
+                        let params = ReplicatedLogletParams {
+                            loglet_id: ReplicatedLogletId::new(loglet_id), // todo: should this go through metadata to ensure global uniqueness?
+                            nodeset: node_set.clone(),
+                            replication: self.config.replication.clone(),
+                            sequencer,
+                            write_set: None,
+                        };
+
+                        updated_plan.insert_loglet(TargetLogletState {
+                            log_id: *log_id,
+                            replication: self.config.replication.clone(),
+                            loglet_id: ReplicatedLogletId::new(loglet_id),
+                            params: LogletParams::from(
+                                params.serialize().expect("can always serialize params"),
+                            ),
+                            segment_index: SegmentIndex::default(),
+                            node_set,
+                            sequencer: Some(sequencer),
+                        });
+
+                        // todo: what if we previously proposed this action? how do we avoid doing this unnecessarily?
+                        effects.push(LogletEffect::AddLogletSegment);
```

Comment:
I guess we could filter it out when trying to write the updated Logs configuration.